classes/Image.js

```diff
@@ -0,0 +1,243 @@
+            img.name = response.rows[i].name;
+            img.link = response.rows[i].link;
+            img.added = response.rows[i].added;
+            images.push(img);
+        }
+        return images;
+    }
+
+    /**
+     * Get a subset list of all images in a paginated manner.
+     * @param perPage {number} The total number of images to respond with per page. If less than or equal to 0, all
+     * images are returned.
+     * @param lastImageIndex {number} The index position of the last image in the list from the last time this method
+     * was called. If lastImageIndex < -1 then this value is defaulted to -1.
+     * @returns {Promise<[Image]>} An array of images.
+     * @throws PostgreSQL error
+     */
+    static async getPaginatedImages(perPage, lastImageIndex) {
+        // Go back to page one if an invalid lastImageIndex is provided.
+        if(lastImageIndex == null || lastImageIndex < -1)
+            lastImageIndex = -1;
+        // Return all images if no item count per page is provided.
+        if(perPage == null || perPage <= 0)
+            return (await this.getAllImages()).slice(lastImageIndex + 1);
+
+        const response = await pool.query('SELECT id, name, link, added FROM images ' +
+            'ORDER BY name ASC, added ASC, id ASC LIMIT $1 OFFSET $2', [perPage, lastImageIndex + 1]);
+        const images = [];
+        for(let i = 0; i < response.rows.length; i++) {
+            const img = new Image(response.rows[i].id);
+            img.name = response.rows[i].name;
+            img.link = response.rows[i].link;
+            img.added = response.rows[i].added;
+            images.push(img);
+        }
+        return images;
+    }
+
+    /**
+     * Upload a new Image to the server, and then create a new image in the database.
+     * @param name {string} Name of the image. Required
+     * @param file {File} File stream of the image to upload. Required
+     * @returns {Promise<Image>} The newly uploaded image
+     * @throws PostgreSQL error
+     */
+    static async uploadImage(name, file) {
+
+        // Enforce file types
+        const splitFileName = file.filename.split('.');
+        const fileType = splitFileName[splitFileName.length - 1];
+        if(!PERMITTED_MIMETYPES.includes(file.mimetype.toLowerCase()) ||
+            !PERMITTED_FILETYPES.includes(fileType.toLowerCase())) {
+            throw new Error("Invalid image format! PNG, JPG, JFIF, JPEG, GIF, and WEBP are permitted.")
+        }
+
+        // Image width & height are set temporarily to 0 but are modified by the metadata processor
+        let width = 0;
+        let height = 0;
+        // Hash instance to get a hash digest of the files contents, which serves as the new file name.
+        const nameHash = crypto.createHash('sha1');
+
+        // Validate & get metadata
+        await new Promise((resolve, reject) => {
+            const readStream = file.createReadStream();
+            readStream.on('data', (chunk) => {
```

Comment:
not a node.js expert, but this seems like it could be prone to a race condition. what guarantees that all of the data gets picked up by both this and the .pipe() below? does this rely on the internals of `sharp()` or `metadataFetch.metadata()` not to do any I/O?

Reply:
I think I understand your concern. If lines 178 to 191 finish before line 175 does, the hash will be incorrect. I'll see if I can find out how the `metadata()` method works, but my assumption was [`stream.on('data', ...)`](https://github.com/rpitv/glimpse-api/pull/4/files/3853115d255bc564721898e042205ef7b4bb1333#diff-47144a3a806c1f2c13c000687b1da54dR174) runs **before** all pipes, and [`metadata()`](https://github.com/rpitv/glimpse-api/pull/4/files/3853115d255bc564721898e042205ef7b4bb1333#diff-47144a3a806c1f2c13c000687b1da54dR178-R179) waits for the [`pipe()`](https://github.com/rpitv/glimpse-api/pull/4/files/3853115d255bc564721898e042205ef7b4bb1333#diff-47144a3a806c1f2c13c000687b1da54dR196) to complete before running the code in [`.then()`](https://github.com/rpitv/glimpse-api/pull/4/files/3853115d255bc564721898e042205ef7b4bb1333#diff-47144a3a806c1f2c13c000687b1da54dR179-R191). Please correct me if I'm misunderstanding your concern though.

Reply:
From the [Sharp docs](https://sharp.readthedocs.io/en/v0.11.0/api/#metadatacallback):
> ### metadata([callback])
> Fast access to image metadata without decoding any compressed image data.
So sounds like it's possible `metadata()` could resolve before the entire image contents are piped through. Could be easily fixed by just creating a separate stream to hash the file name, or just come up with a different naming scheme entirely.